Bearded.UI/Events/EventRouter.cs

```diff
@@ -42,12 +36,55 @@ public static EventPropagationPath FindPropagationPath(IControlParent root, Cont
                     return EventPropagationPath.Empty;
                 }
                 path.Add(parentAsControl);
-                
+
                 parent = parentAsControl.Parent;
             }
 
             path.Reverse();
             return new EventPropagationPath(path.AsReadOnly());
         }
+
+        private static IEnumerable<Control> findPropagationPathEnumerable(
+            IControlParent root, PropagationTest propagationTest)
+        {
+            if (root == null) yield break;
+
+            foreach (var child in root.Children.Reverse())
+            {
+                var outcome = propagationTest(child);
+                switch (outcome)
+                {
+                    case PropagationTestOutcome.Failure:
+                        continue;
+                    case PropagationTestOutcome.PassThrough:
+                        if (!(child is IControlParent childAsParent) || childAsParent.Children.Count == 0)
+                        {
+                            continue;
+                        }
+                        var potentialPath = findPropagationPathEnumerable(childAsParent, propagationTest).ToList();
+                        if (potentialPath.Count == 0
+                            || propagationTest(potentialPath.Last()) == PropagationTestOutcome.PassThrough)
+                        {
+                            continue;
+                        }
+                        yield return child;
+                        foreach (var descendant in potentialPath)
+                        {
+                            yield return descendant;
+                        }
+                        break;
```

Comment:
Is there a reason to `break` here, while you explicitly `continue` above several times? If not, I'd like this to be a `continue` as well to be explicit and consistent and to differentiate it from the `yield break` below more. It's otherwise easy to mistake this `break` for referring to the outer loop, which it does not (though ironically, the `continue` statements do, still they are less ambiguous).

Reply:
I think you found a bug here, and this should actually be a yield break. However, I want to try if rewriting this using a stack is clearer, though I think it will be a bit annoying given the fact you need to somehow backtrack if you only have pass-through levels.

Reply:
I thought it was a bug, but then I stepped through and it made sense, and now I'm not sure. :D 

RE stack: You can either backtrack element by element, or count how many passthroughs you have. Perhaps using a list and just storing the last index of a non-passthrough is easier actually. It's still basically a stack, but you can remove multiple elements easier... Oh, except if you go back up the stack, you list which of the siblings you were looking at, at that step. That might actually be a major issue with this idea...

Reply:
Of course, you could pass the list into a recursive function call.

Reply:
This is what I ended up doing, and I think it did indeed make the algorithm much clearer and performant. Please take a look and let me know if this makes sense to you. Thanks for the feedback!